lib/caching-precompiler.js

```diff
@@ -1,142 +1,180 @@
 'use strict';
-var fs = require('fs');
 var path = require('path');
+var fs = require('fs');
+var generateMapFileComment = require('convert-source-map').generateMapFileComment;
+var convertSourceMap = require('convert-source-map');
 var cachingTransform = require('caching-transform');
-var md5Hex = require('md5-hex');
-var stripBom = require('strip-bom');
 var objectAssign = require('object-assign');
-var generateMapFileComment = require('convert-source-map').generateMapFileComment;
-
-module.exports = CachingPrecompiler;
+var stripBom = require('strip-bom');
+var md5Hex = require('md5-hex');
+var enhanceAssert = require('./enhance-assert');
 
-function CachingPrecompiler(cacheDir, babelConfig) {
+function CachingPrecompiler(cacheDirPath, babelConfig) {
 	if (!(this instanceof CachingPrecompiler)) {
 		throw new TypeError('Class constructor CachingPrecompiler cannot be invoked without \'new\'');
 	}
 
-	this.cacheDir = cacheDir;
-	this.filenameToHash = {};
-	this.transform = this._createTransform(babelConfig);
+	this.babelConfig = babelConfig;
+	this.cacheDirPath = cacheDirPath;
+	this.fileHashes = {};
+
+	Object.keys(CachingPrecompiler.prototype).forEach(function (name) {
+		this[name] = this[name].bind(this);
+	}, this);
+
+	this.transform = this._createTransform();
 }
 
-CachingPrecompiler.prototype._factory = function (babelConfig, cacheDir) {
-	// This factory method is only called once per process, and only as needed, to defer loading expensive dependencies.
-	var babel = require('babel-core');
-	var convertSourceMap = require('convert-source-map');
-	var presetStage2 = require('babel-preset-stage-2');
-	var presetES2015 = require('babel-preset-es2015');
-	var transformRuntime = require('babel-plugin-transform-runtime');
-
-	var rewriteRuntime = this._createRewritePlugin();
-	var powerAssert = this._createEspowerPlugin(babel);
-
-	function buildOptions(filename, code) {
-		// Extract existing source maps from the code.
-		var sourceMap = convertSourceMap.fromSource(code) || convertSourceMap.fromMapFileSource(code, path.dirname(filename));
-
-		var options = {babelrc: false};
-
-		if (!babelConfig || babelConfig === 'default') {
-			objectAssign(options, {presets: [presetStage2, presetES2015]});
-		} else if (babelConfig === 'inherit') {
-			objectAssign(options, {babelrc: true});
-		} else {
-			objectAssign(options, babelConfig);
-		}
-
-		objectAssign(options, {
-			inputSourceMap: sourceMap && sourceMap.toObject(),
-			filename: filename,
-			sourceMaps: true,
-			ast: false
-		});
-
-		options.plugins = options.plugins || [];
-		options.plugins.push(powerAssert, transformRuntime, rewriteRuntime);
-
-		return options;
+module.exports = CachingPrecompiler;
+
+CachingPrecompiler.prototype._factory = function () {
+	this.init();
+
+	return this._transform;
+};
+
+CachingPrecompiler.prototype.init = function () {
```

Comment:
Maybe inline this logic in `_factory`? `CachingPrecompiler` is really a wrapper around `caching-transform`. `init()` implies it needs to be explicitly initialized which isn't the case.